src/Lykke.AlgoStore.CSharp.AlgoTemplate.Services/Functions/ADX/DirectionalMovementIndexPlusFunction.cs

```diff
@@ -0,0 +1,167 @@
+﻿using Lykke.AlgoStore.CSharp.Algo.Core.Candles;
+using Lykke.AlgoStore.CSharp.Algo.Core.Functions;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Lykke.AlgoStore.CSharp.AlgoTemplate.Services.Functions.ADX
+{
+    public class DirectionalMovementIndexPlusFunction : IFunction
+    {
+        private readonly int _period;
+        private readonly bool _isAverageTrueRangeSet = false;
+        private int _samples { get; set; }
+
+        private DMIParameters _functionParams = new DMIParameters();
+        public FunctionParamsBase FunctionParameters => _functionParams;
+
+        private double? DirectionalMovementIndexPlus { get; set; }
+        private double? SmoothedDirectionalMovementPlus { get; set; }
+        private double? PreviousSmoothedDirectionalMovementPlus { get; set; }
+
+        private Queue<double> DirectionalMovementPluses { get; set; }
+        private double DirectionalMovementPlus { get; set; }
+
+        /// <summary>
+        /// ATR
+        /// </summary>
+        public double? AverageTrueRange { get; set; }
+
+        private Candle _previousInput;
+        private ATRFunction ATRFunction { get; set; }
+
+        public DirectionalMovementIndexPlusFunction(DMIParameters dmiParameters)
+        {
+            _period = dmiParameters.Period;
+            _functionParams = dmiParameters;
+            DirectionalMovementPluses = _functionParams.Period == 0 ? new Queue<double>() : new Queue<double>(_functionParams.Period);
+            _isAverageTrueRangeSet = _functionParams.IsAverageTrueRangeSet;
+            AverageTrueRange = 0.0d;
+            ATRFunction = new ATRFunction(new AtrParameters() { Period = _period });
+            _samples = 0;
+        }
+
+        public double? WarmUp(IList<Candle> values)
+        {
+            if (values == null)
+                throw new ArgumentException();
+
+            foreach (var value in values)
+            {
+                _samples++;
+
+                if (!_isAverageTrueRangeSet)
+                {
+                    AverageTrueRange = ATRFunction.AddNewValue(value);
+                }
+
+                DirectionalMovementPlus = ComputePositiveDirectionalMovement(value);
+
+                if (values.IndexOf(value) > 0 && !IsReady)
+                {
+                    DirectionalMovementPluses.Enqueue(DirectionalMovementPlus);
+                }
+
+                SmoothedDirectionalMovementPlus = ComputeSmoothedDirectionalMovementPlus();
+                DirectionalMovementIndexPlus = ComputePositiveDirectionalIndex();
+
+                _previousInput = value;
+            }
+
+            return DirectionalMovementIndexPlus;
+        }
+
```

Comment:
Can move it to top and use expression-bodied property "public bool IsReady => _samples > _period + 1;"